models/inventory/StockManager.ts

```diff
@@ -126,6 +127,58 @@ export class StockManager {
     throw new ValidationError(t`Both From and To Location cannot be undefined`);
   }
 
+  async #validateBatchWiseStockAvailability(details: SMIDetails) {
+    /*
+     * Checks if hasBatchNumber is enabled in the item
+     * If user has not entered batchNumber raises a ValidationError
+     * If entered quantity is greater than the available quantity in the batch, raises a ValidationError
+     */
+
+    if (!details.fromLocation) {
+      return;
+    }
+
+    const ifItemHasBatchNumber = await this.fyo.getValue(
+      'Item',
+      details.item,
+      'hasBatchNumber'
+    );
+    const date = details.date.toISOString();
+    const formattedDate = this.fyo.format(details.date, 'Datetime');
+
+    if (ifItemHasBatchNumber) {
+      if (!details.batchNumber) {
+        throw new ValidationError(
+          t`Please enter Batch Number for ${details.item}`
+        );
+      }
+
+      const itemsInBatch =
+        (await this.fyo.db.getStockQuantity(
+          details.item,
+          details.fromLocation,
+          undefined,
+          date,
+          details.batchNumber
+        )) ?? 0;
+
+      if (details.quantity > itemsInBatch) {
```

Comment:
Same nitpick here; this function can be flattened to have only one sub-nested block by using early returns.

Reply:
corrected